File: shareService/webapp/src/main/java/uber/megashare/controller/GenericEditController.java
/**
 * Copyright (C) 2011 alex <alex@0x08.tk>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *         http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package uber.megashare.controller;

import java.util.Locale;
import javax.servlet.http.HttpServletRequest;
import javax.validation.Valid;
import org.springframework.ui.Model;
import org.springframework.validation.BindingResult;
import org.springframework.validation.FieldError;
import org.springframework.web.bind.annotation.ModelAttribute;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestMethod;
import org.springframework.web.bind.annotation.RequestParam;
import static uber.megashare.controller.EditConstants.MODEL_KEY;
import uber.megashare.model.Struct;
import uber.megashare.service.StructManager;

/**
 *
 * @author alex
 */
public abstract class GenericEditController<T extends Struct> extends GenericController<T> implements EditConstants {

    /**
     *
     */
    private static final long serialVersionUID = 3844308254206235808L;

    protected GenericEditController(StructManager<T> manager) {
        super(manager);
    }

    /**
     * get new model instance associated with this controller see implementation
     * in child controllers
     *
     * @return
     */
    protected abstract T getNewModelInstance();

    protected void putModel(T obj, Model model) {

        if (!obj.isBlank()) {
            if (manager.exists(obj.getId() + 1)) {
                model.addAttribute("availableNext", 1);
            }

            if (manager.exists(obj.getId() - 1)) {
                model.addAttribute("availablePrev", 1);
            }
        }
    }

    protected abstract boolean checkAccess(T obj, Model model);

    @RequestMapping(value = "/{id:[0-9]+}", method = RequestMethod.GET)
    public String getByPath(@PathVariable("id") Long id, Model model, Locale locale) {
      return view(id,model,locale);
    }

    @RequestMapping(value = EDIT_ACTION + "/{id:[0-9]+}", method = RequestMethod.GET)
    public String editByPath(@PathVariable("id") Long id, Model model, Locale locale) {
        return edit(model, id, locale);
    }

    @RequestMapping(value = VIEW_ACTION, method = RequestMethod.GET)
    public String view(@RequestParam(required = false) Long id, Model model, Locale locale) {

        String result = edit(model, id, locale);
        return result.equals(editPage) ? viewPage : result;
    }

    @RequestMapping(value = EDIT_ACTION, method = RequestMethod.GET)
    public String edit(Model model, @RequestParam(required = false) Long id, Locale locale) {

        if (id == null) {
            return listPage;
        }

        T omodel = manager.getFull(id);

        if (omodel == null) {
            return page404;
        }

        if (!checkAccess(omodel, model)) {
            return page403;
        }

        model.addAttribute(MODEL_KEY, omodel);

        putModel(omodel, model);

        return editPage;
    }

    @RequestMapping(value = NEW_ACTION, method = RequestMethod.GET)
    public String createNew(Model model) {

        /**
         * allow create only for authorized users
         */
        if (!isCurrentUserLoggedIn()) {
            addMessageDenied(model);
            return listPage;
        }
        model.addAttribute(MODEL_KEY, getNewModelInstance());

        return editPage;
    }

    @RequestMapping(value = SAVE_ACTION, method = RequestMethod.POST)
    public String save(@RequestParam(required = false) String cancel,
            @Valid
            @ModelAttribute(MODEL_KEY) T b,
            BindingResult result, Model model, HttpServletRequest request) {

        /**
         * allow upload only for authorized users
         */
        if (!isCurrentUserLoggedIn()) {
            addMessageDenied(model);
            return editPage;
        }

        if (cancel != null) {
            addMessageCancelled(model);
            return listPage;
        }

        if (result.hasErrors()) {
            if (getLogger().isDebugEnabled()) {
                debugPrintRejected(result);
            }
            return editPage;
        }

        T r = manager.save(b);


        /**
         * set id from create this is needed to correct model's id postback when
         * validation fails
         */
        if (b.isBlank()) {
            b.setId(r.getId());
        }

        resetPagingList(request);

        addMessageSuccess(model);
        return listPage;
    }

    @RequestMapping(value = DELETE_ACTION, method = {RequestMethod.GET, RequestMethod.POST})
    public String delete(@RequestParam(required = true) Long id,
            Model model, HttpServletRequest request) {
        manager.remove(id);
        resetPagingList(request);
        return listPage;
    }

    protected T get(Long id) {
        getLogger().debug("get by id=" + id);
        return manager.getFull(id);
    }

    protected void debugPrintRejected(BindingResult result) {
        getLogger().debug("form has errors " + result.getErrorCount());
        for (FieldError f : result.getFieldErrors()) {
            getLogger().debug("field=" + f.getField() + ",rejected value=" + f.getRejectedValue() + ",message=" + f.getDefaultMessage());
        }
    }
}

File: uberPaste/src/main/scala/uber/paste/controller/GenericEditController.scala
/*
 * Copyright 2011 Ubersoft, LLC.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package uber.paste.controller

import uber.paste.model.Struct
import org.springframework.http.HttpStatus
import org.springframework.ui.Model
import org.springframework.validation.BindingResult
import org.springframework.web.bind.annotation._
import javax.validation.Valid
import java.util.Locale
import org.springframework.web.servlet.mvc.support.RedirectAttributes

object GenericEditController {

  final val EDIT_ACTION = "/edit"
  final val VIEW_ACTION = "/view"
  final val SAVE_ACTION= "/save"
  final val NEW_ACTION= "/new"
  final val DELETE_ACTION= "/delete"

}

abstract class GenericEditController[T <: Struct ] extends StructController[T] {    
  
  
    protected def getNewModelInstance():T

    protected def fillEditModel(obj:T,rev:Long,model:Model,locale:Locale) {
      model.addAttribute(GenericController.MODEL_KEY, obj)
    }

    def loadModel(@RequestParam(required = false) id:java.lang.Long):T = {
      logger.debug("_new request id="+id)

        return if (id != null)
          manager.getFull(id)
        else
          null.asInstanceOf[T]
    }

  @RequestMapping(value = Array(GenericEditController.NEW_ACTION), method = Array(RequestMethod.GET))
  @ResponseStatus(HttpStatus.CREATED)
  def createNew(model:Model,locale:Locale):String= {

    fillEditModel(getNewModelInstance(),-1,model,locale)

    return editPage
  }


  @RequestMapping(value = Array("/edit/{id:[0-9]+}"), method = Array(RequestMethod.GET))
  def editWithId(model:Model,@PathVariable("id") id:Long,locale:Locale):String= {

      fillEditModel(loadModel(id),-1,model,locale)

    return editPage
  }


    @RequestMapping(value = Array(GenericEditController.SAVE_ACTION), method = Array(RequestMethod.POST))
    def save(@RequestParam(required = false) cancel:String,
            @Valid @ModelAttribute(GenericController.MODEL_KEY) b:T,
            result:BindingResult, model:Model,locale:Locale,
            redirectAttributes:RedirectAttributes):String = {

        if (cancel != null) {
            redirectAttributes.addFlashAttribute("statusMessageKey", "action.cancelled")
            return listPage
        }

        if (result.hasErrors()) {
              logger.debug("form has errors " + result.getErrorCount())
              
             fillEditModel(b,-1,model,locale)
           /* for ( f:FieldError : result.getFieldErrors()) {
                getLogger().debug("field=" + f.getField() + ",rejected value=" + f.getRejectedValue()+",message="+f.getDefaultMessage());
            }*/
            return editPage
        }

        val r:T = manager.save(b)

        // set id from create
        if (b.isBlank) {
            b.setId(r.getId())
        }

      redirectAttributes.addFlashAttribute("statusMessageKey", "action.success")

      return listPage
    }

   

    @RequestMapping(value = Array(GenericEditController.DELETE_ACTION), method = Array(RequestMethod.GET,RequestMethod.POST))
    def delete(@RequestParam(required = false) id:Long,
            model:Model):String = {
        manager.remove(id);
        return listPage;
    }



  @RequestMapping(value = Array("/{id:[0-9]+}"), method = Array(RequestMethod.GET))
  def getByPath(@PathVariable("id") id:java.lang.Long,model:Model,locale:Locale):String = {

    val m = loadModel(id)

    if (m==null)
      return page404

    model.addAttribute(GenericController.MODEL_KEY, m)

    fillEditModel(m,-1,model,locale)

    return viewPage
  }


  @RequestMapping(value = Array("/xml/{id:[0-9]+}"), method = Array(RequestMethod.GET))
  @ResponseBody
  def getBody(@PathVariable("id") id:java.lang.Long,model:Model,locale:Locale):T = {
    return loadModel(id);
}


}
